// dataverse-utils/lib/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.logger = exports.icons = void 0;
const kleur_1 = __importDefault(require("kleur"));
const figures_1 = __importDefault(require("figures"));
const isTest = process.env.JEST_WORKER_ID !== undefined;
exports.icons = {
    done: kleur_1.default.green(figures_1.default.tick),
    info: kleur_1.default.cyan(figures_1.default.pointer),
    error: kleur_1.default.red(figures_1.default.cross),
    warn: kleur_1.default.yellow(figures_1.default.warning)
};
exports.logger = {
    info(...args) {
        if (!isTest) {
            console.info(exports.icons.info, ...args);
        }
    },
    warn(...args) {
        if (!isTest) {
            console.warn(exports.icons.warn, ...args);
        }
    },
    error(...args) {
        if (!isTest) {
            console.error(exports.icons.error, ...args);
        }
    },
    done(...args) {
        if (!isTest) {
            console.info(exports.icons.done, ...args);
        }
    },
};
